app/models/puppetclass.rb

```diff
@@ -45,30 +46,133 @@ class Puppetclass < ActiveRecord::Base
+              puts "Rewriting..." if verbose
+              cmd = "ruby -p -i -e 'rex=%r{#{root}};$_.gsub!(rex,\"\")' #{files}"
+              puts cmd if debug
+              sh cmd
+              # Now relocate the files/* files to match the rewritten url
+              mv Dir.glob("#{out}/files/#{root}/*"), "#{out}/files", :verbose => verbose
+            end
+          else
+            logger.warn "Failed to process puppetdocs for #{out} while executing #{cmd}"
+            puts "Failed to process puppetdocs for #{out} while executing #{cmd}"
+          end
+        end
+        puts
       end
     end
-    hash
-  end
 
-  # For API v2 - eliminate node :puppetclass for each object. returns a hash containing modules and associated classes
-  def self.classes2hash_v2(classes)
-    hash = {}
-    classes.each do |klass|
-      if (mod = klass.module_name)
-        hash[mod] ||= []
-        hash[mod] << {:id => klass.id, :name => klass.name, :created_at => klass.created_at, :updated_at => klass.updated_at}
+    # Optionally creates a copy of the current puppet modules and sanitizes it.
+    # If your 'live' manifests and modules can be parsed by puppetdoc
+    # then you do not need to do this step. (Unfortunately some sites have circular
+    # symlinks which have to be removed.)
+    # If the executable Rails,root/script/rdoc_prepare_script exists then it is run
+    # and passed a list of all directory paths in all environments.
+    # It should return the directory into which it has copied the cleaned modules"
+    def prepare_rdoc(root)
+      debug, verbose = false, false
+
+      prepare_script = Pathname.new(Rails.root) + "script/rdoc_prepare_script.rb"
+      if prepare_script.executable?
+        dirs = Environment.puppetEnvs.values.join(":").split(":").uniq.sort.join(" ")
+        puts "Running #{prepare_script} #{dirs}" if debug
+        location = `#{prepare_script} #{dirs}`
+        if $CHILD_STATUS == 0
+          root = location.chomp
+          puts "Relocated modules to #{root}" if verbose
+        end
+      else
+        puts "No executable #{prepare_script} found so using the uncopied module sources" if verbose
       end
+      root
+    end
+
+    def search_by_host(key, operator, value)
+      conditions = sanitize_sql_for_conditions(["hosts.name #{operator} ?", value_to_sql(operator, value)])
+      direct     = Puppetclass.joins(:hosts).where(conditions).uniq.pluck('puppetclasses.id')
+      hostgroup  = Hostgroup.joins(:hosts).where(conditions).first
+      indirect   = hostgroup.blank? ? [] : HostgroupClass.where(:hostgroup_id => hostgroup.path_ids).uniq.pluck('puppetclass_id')
+      return { :conditions => "1=0" } if direct.blank? && indirect.blank?
+
+      puppet_classes = (direct + indirect).uniq
+      { :conditions => "puppetclasses.id IN(#{puppet_classes.join(',')})" }
+    end
+
+    def permitted_attributes
+      [:name, { :hostgroup_ids => [], :host_ids => [], :environment_ids => [] } ]
```

Comment:
How does this work with `_names` too, does our concern rewrite this as `_ids`, or do we need to whitelist both?